File: optimizer-api/optimizer.py
"""
CarbonGPT Prompt Optimizer – FastAPI Edition
=============================================
A comprehensive prompt optimizer that removes filler phrases,
hedging language, politeness markers, and off-topic asides
while preserving the core meaning of the prompt.

This module mirrors the logic from the Flask-based optimizer
(carbongpt-backend/python-optimizer/optimizer.py) and is used
by the FastAPI service in main.py.
"""

import re
from typing import List, Tuple


# ─── Comprehensive phrase patterns to remove ────────────────────────

PHRASE_PATTERNS = [
    # Politeness markers
    r"\bplease\b",
    r"\bpls\b",
    r"\bplease do\b",
    r"\bkindly\b",
    r"\bkindly do\b",
    r"\bthanks\b",
    r"\bthank you\b",
    r"\bthank you very much\b",
    r"\bthank you so much\b",
    r"\bthank you kindly\b",
    r"\bthanks a lot\b",
    r"\bthanks a bunch\b",
    r"\bthanks a million\b",
    r"\bthanks so much\b",
    r"\bthank you for your time\b",
    r"\bthank you for your patience\b",
    r"\bthank you for your consideration\b",
    r"\bthank you for your help\b",
    r"\bthank you for everything\b",
    r"\bmuch obliged\b",
    r"\bi appreciate it\b",
    r"\bi appreciate your help\b",
    r"\bi'm grateful\b",
    r"\bwarm regards\b",
    r"\bkind regards\b",
    r"\bbest regards\b",
    r"\bregards\b",
    r"\brespectfully\b",
    r"\bwith all due respect\b",
    r"\bwould you kindly\b",
    r"\bcould you kindly\b",
    r"\bexcuse me\b",
    r"\bpardon me\b",
    r"\bsorry\b",
    r"\bi'm sorry\b",
    r"\bmy apologies\b",
    r"\bi apologize\b",
    r"\bif you don't mind\b",
    r"\bno offense\b",
    r"\bno disrespect\b",
    r"\bwith permission\b",
    r"\bwith your permission\b",
    r"\bfor your convenience\b",
    r"\bfor your information\b",
    r"\bfyi\b",
    r"\bat your service\b",
    r"\bat your disposal\b",
    r"\bmy pleasure\b",

    # Question softeners
    r"\bcould you\b",
    r"\bwould you\b",
    r"\bwould you mind\b",
    r"\bcan you\b",
    r"\bcan you please\b",
    r"\bcould you please\b",
    r"\bwould you please\b",
    r"\bwould you be so kind\b",
    r"\bcould you be so kind\b",
    r"\bi was wondering if\b",
    r"\bi was wondering whether\b",
    r"\bwould you be willing to\b",
    r"\bwould you mind if\b",
    r"\bcould you possibly\b",
    r"\bwould you possibly\b",
    r"\bif you would\b",
    r"\bif possible\b",
    r"\bif you can\b",
    r"\bif you could\b",
    r"\bif it's not too much trouble\b",
    r"\bif you have time\b",
    r"\bif you have a moment\b",
    r"\bwhen you get a chance\b",
    r"\bat your convenience\b",
    r"\bat your earliest convenience\b",
    r"\bwhenever you can\b",
    r"\bdo you think you could\b",
    r"\bwould it be possible\b",
    r"\bwould it be possible to\b",
    r"\bis it possible\b",
    r"\bis it possible to\b",
    r"\bby any chance\b",
    r"\bmay i ask\b",
    r"\bcan i ask\b",
    r"\bi would be grateful if\b",
    r"\bi would appreciate it if\b",
    r"\bi would greatly appreciate it if\b",

    # Hedging and uncertainty
    r"\bmaybe\b",
    r"\bperhaps\b",
    r"\bpossibly\b",
    r"\bprobably\b",
    r"\bseems\b",
    r"\bseems to\b",
    r"\bappears\b",
    r"\bappears to\b",
    r"\bi think\b",
    r"\bi believe\b",
    r"\bi suspect\b",
    r"\bi suppose\b",
    r"\bi guess\b",
    r"\bin my opinion\b",
    r"\bin my view\b",
    r"\bin my humble opinion\b",
    r"\bif you ask me\b",
    r"\bfrom my perspective\b",
    r"\bas far as i know\b",
    r"\bas far as i can tell\b",
    r"\bto my knowledge\b",
    r"\bif i recall correctly\b",
    r"\bif i'm not mistaken\b",
    r"\bcorrect me if i'm wrong\b",
    r"\bi could be wrong\b",
    r"\bi'm not sure\b",
    r"\bi'm uncertain\b",
    r"\bseems like\b",
    r"\blooks like\b",
    r"\bsounds like\b",
    r"\bfeels like\b",
    r"\bfor whatever reason\b",
    r"\bfor some reason\b",
    r"\ballegedly\b",
    r"\breportedly\b",
    r"\bsort of\b",
    r"\bkind of\b",
    r"\bpartly\b",
    r"\bpartially\b",
    r"\bto some extent\b",
    r"\bto a degree\b",
    r"\bmore or less\b",
    r"\bso to speak\b",
    r"\bin a manner of speaking\b",
    r"\bif you will\b",
    r"\bas it were\b",
    r"\bapparently\b",
    r"\bsomewhat\b",
    r"\brather\b",
    r"\broughly\b",
    r"\bapproximately\b",

    # Discourse markers
    r"\byou know\b",
    r"\blike\b",
    r"\banyway\b",
    r"\banyhow\b",
    r"\bby the way\b",
    r"\bat any rate\b",
    r"\bon second thought\b",
    r"\bwell\b",
    r"\bso\b",
    r"\blook\b",
    r"\blisten\b",
    r"\bright\b",
    r"\bokay\b",
    r"\byou see\b",
    r"\bi mean\b",
    r"\bas i was saying\b",
    r"\bin any case\b",
    r"\bfurthermore\b",
    r"\bmoreover\b",
    r"\bbesides\b",
    r"\bhowever\b",
    r"\bnevertheless\b",
    r"\btherefore\b",
    r"\bthus\b",
    r"\bhence\b",
    r"\baccordingly\b",
    r"\badmittedly\b",
    r"\bas a matter of fact\b",
    r"\bbelieve it or not\b",
    r"\bcome to think of it\b",
    r"\bnow that i think about it\b",
    r"\bthat reminds me\b",
    r"\bas it turns out\b",
    r"\bit turns out\b",
    r"\bincidentally\b",
    r"\bthe thing is\b",
    r"\bthe point is\b",
    r"\bhere's the thing\b",
    r"\bfor the record\b",
    r"\bi'm not gonna lie\b",
    r"\bfrankly\b",
    r"\bto be frank\b",
    r"\bto be honest\b",
    r"\bif i'm being honest\b",
    r"\bin effect\b",
    r"\beffectively\b",
    r"\bin essence\b",
    r"\bin short\b",
    r"\bin brief\b",
    r"\bin a nutshell\b",

    # Obvious fillers
    r"\bjust\b",
    r"\bquickly\b",
    r"\bmerely\b",
    r"\bonly\b",
    r"\bpurely\b",
    r"\bdirectly\b",
    r"\bactually\b",
    r"\bbasically\b",
    r"\bliterally\b",
    r"\bhonestly\b",
    r"\bseriously\b",
    r"\bcompletely\b",
    r"\bentirely\b",
    r"\btotally\b",
    r"\babsolutely\b",
    r"\bdefinitely\b",
    r"\bundoubtedly\b",
    r"\bfor sure\b",
    r"\bfor certain\b",
]

# Structural words to never remove
PROTECTED_WORDS = {
    "not", "no", "never", "nor", "or", "and", "but", "because", "since",
    "although", "though", "while", "if", "when", "where", "why", "how",
    "who", "what", "which", "whoever", "whatever", "whichever",
    "they", "them", "their", "you", "your", "we", "us", "our", "i", "me", "my",
    "he", "him", "his", "she", "her", "is", "are", "was", "were", "be", "been", "being",
    "have", "has", "had", "do", "does", "did", "will", "would", "should", "could", "can",
    "may", "might", "must", "shall", "all", "both", "each", "either",
    "any", "every", "as", "than", "then", "from", "to", "in", "on", "at", "by", "of",
    "with", "without", "about", "into", "through", "during", "before", "after", "above",
    "below", "up", "down", "that", "which", "this", "these", "those", "the", "a", "an",
}


# ─── Utility functions ──────────────────────────────────────────────

def normalize_text(text: str) -> str:
    """Normalize whitespace and clean up text."""
    text = text.strip()
    text = re.sub(r"\s+", " ", text)
    return text


def detect_vague_phrases(text: str) -> List[str]:
    """
    Detect common vague/filler phrases NOT in the explicit list.
    Catches unknown fillers by pattern characteristics.
    """
    vague_patterns = [
        r"\b(?:after a while|in a bit|earlier|later|soon|the other day)\b",
        r"\b(?:a bunch|a ton|loads of|tons of|quite a few|numerous)\b",
        r"\b(?:and stuff|and things|or whatever|or something|or anything|and whatnot|or whatnot)\b",
        r"\b(?:i guess you could say|how should i say|let me think|hmm|honestly)\b",
        r"\b(?:i'm no expert|i'm not sure but|correct me if wrong|not an expert but)\b",
        r"\b(?:moving on|to make long story short|bottom line|in conclusion)\b",
        r"\b(?:seriously|no kidding|you know what|get this)\b",
        r"\b(?:randomly|totally|completely by chance)\b",
    ]

    found_phrases = []
    for pattern in vague_patterns:
        matches = re.finditer(pattern, text, re.IGNORECASE)
        for match in matches:
            found_phrases.append(match.group())

    return list(set(found_phrases))


def remove_redundant_qualifiers(text: str) -> str:
    """Remove redundant qualifiers and clean up formatting."""
    # Repeated qualifiers
    text = re.sub(r"\b(very|really|so|quite|rather|extremely)\s+\1+\b", r"\1", text, flags=re.IGNORECASE)

    # Multiple punctuation collapsed
    text = re.sub(r"\?{2,}", "?", text)
    text = re.sub(r"!{2,}", "!", text)
    text = re.sub(r"\.{2,}", ".", text)
    text = re.sub(r"…{2,}", "…", text)

    # Empty parentheses/brackets
    text = re.sub(r"\(\s*\)", "", text)
    text = re.sub(r"\[\s*\]", "", text)

    # Stray commas
    text = re.sub(r"^\s*,", "", text)
    text = re.sub(r",\s*,", ",", text)

    # Double spaces
    text = re.sub(r"\s{2,}", " ", text)
    return text


def remove_off_topic_asides(text: str, remove_asides: bool = True) -> Tuple[str, List[str]]:
    """
    Identify and optionally remove off-topic parenthetical content.
    Returns (cleaned_text, list_of_removed_asides).
    """
    aside_keywords = [
        "ignore", "forget", "skip", "don't mind", "disregard",
        "not sure if", "not that", "not important", "random",
        "by the way", "tangent", "aside", "never mind"
    ]

    asides = re.findall(r"\([^)]*\)", text)
    removed_asides = []

    for aside in asides:
        if any(keyword in aside.lower() for keyword in aside_keywords):
            removed_asides.append(aside)
            text = text.replace(aside, "", 1)

    return normalize_text(text), removed_asides


def remove_phrase_patterns(text: str) -> str:
    """Remove filler phrases while preserving structure."""
    for pattern in PHRASE_PATTERNS:
        text = re.sub(pattern, " ", text, flags=re.IGNORECASE)
    return normalize_text(text)


def smart_deduplicate(text: str, max_repeats: int = 1) -> str:
    """Remove repeated adjacent words and phrases."""
    words = text.split()
    if not words:
        return text

    result: List[str] = []
    repeat_count = 0
    last_word = words[0]
    result.append(last_word)

    for word in words[1:]:
        if word.lower() == last_word.lower():
            repeat_count += 1
            if repeat_count < max_repeats:
                result.append(word)
        else:
            last_word = word
            repeat_count = 0
            result.append(word)

    words = result

    # Collapse repeated adjacent phrases (length 2–3)
    for phrase_len in (3, 2):
        i = 0
        while i + 2 * phrase_len <= len(words):
            if [w.lower() for w in words[i:i + phrase_len]] == [w.lower() for w in words[i + phrase_len:i + 2 * phrase_len]]:
                del words[i + phrase_len:i + 2 * phrase_len]
            else:
                i += 1

    return " ".join(words)


# ─── Main optimization functions ────────────────────────────────────

def optimize_prompt(prompt: str, aggressive: bool = False, remove_asides: bool = True) -> str:
    """
    Optimize a prompt by removing fillers while preserving meaning.

    Args:
        prompt: The prompt text to optimize
        aggressive: If True, also remove light stopwords (experimental)
        remove_asides: If True, remove off-topic parenthetical asides

    Returns:
        Optimized prompt string
    """
    prompt = normalize_text(prompt)

    # Step 1: Remove off-topic asides
    prompt, _ = remove_off_topic_asides(prompt, remove_asides=remove_asides)

    # Step 2: Remove explicitly listed filler phrases
    prompt = remove_phrase_patterns(prompt)

    # Step 3: Remove vague/unknown filler phrases detected by pattern
    for vague_phrase in detect_vague_phrases(prompt):
        prompt = re.sub(re.escape(vague_phrase), " ", prompt, flags=re.IGNORECASE)

    # Step 4: Remove redundant qualifiers and clean up formatting
    prompt = remove_redundant_qualifiers(prompt)

    # Step 5: Smart deduplication
    prompt = smart_deduplicate(prompt, max_repeats=1)

    # Final cleanup
    prompt = normalize_text(prompt)
    return prompt


def batch_optimize(prompts: List[str], aggressive: bool = False, remove_asides: bool = True) -> List[str]:
    """Optimize multiple prompts."""
    return [optimize_prompt(p, aggressive=aggressive, remove_asides=remove_asides) for p in prompts]


def analyze_prompt(prompt: str, remove_asides: bool = True, aggressive: bool = False) -> dict:
    """
    Analyze prompt and return detailed statistics including unknown phrases detected.
    """
    original_words = len(prompt.split())

    # Detect unknown vague phrases BEFORE removal
    vague_phrases = detect_vague_phrases(prompt)

    # Detect off-topic asides BEFORE removal
    _, removed_asides = remove_off_topic_asides(prompt, remove_asides=False)

    # Optimize
    optimized = optimize_prompt(prompt, remove_asides=remove_asides, aggressive=aggressive)
    optimized_words = len(optimized.split())
    reduction = ((original_words - optimized_words) / original_words * 100) if original_words > 0 else 0

    return {
        "original_word_count": original_words,
        "optimized_word_count": optimized_words,
        "reduction_percentage": round(reduction, 2),
        "original": prompt,
        "optimized": optimized,
        "unknown_vague_phrases_detected": vague_phrases,
        "off_topic_asides": removed_asides,
    }

File: carbongpt-backend/utils/optimizer.js
/**
 * CarbonGPT Prompt Optimizer (Full JS Port)
 * ==========================================
 * Complete port of the Python optimizer — 600+ filler phrase patterns,
 * vague phrase detection, redundant qualifier removal, and smart deduplication.
 *
 * Runs entirely inside Node.js. No Flask or FastAPI needed.
 *
 * Equivalent to: python-optimizer/optimizer.py
 */

// ─── Phrase Patterns ───────────────────────────────────────────────────────

const PHRASE_PATTERNS = [
  // Politeness markers
  "please", "pls", "please do", "kindly", "kindly do",
  "thanks", "thank you", "thank you very much", "thank you so much",
  "thank you kindly", "thanks a lot", "thanks a bunch", "thanks a million",
  "thanks so much", "thank you for your time", "thank you for your patience",
  "thank you for your consideration", "thank you for your help",
  "thank you for everything", "thank you so very much", "thank you ever so much",
  "much obliged", "i appreciate it", "i appreciate your help",
  "we appreciate it", "we appreciate your help", "i'm grateful", "we're grateful",
  "sincere thanks", "with thanks", "with my thanks", "with sincere thanks",
  "warm thanks", "warm regards", "kind regards", "kindest regards",
  "best regards", "regards", "respectfully", "respectfully yours",
  "with respect", "with all due respect", "if i may be so bold",
  "if you would be so kind", "if you would be so gracious",
  "if you would be willing", "if you would be able", "if you would oblige",
  "would you kindly", "would you so kindly", "would you be so gracious",
  "would you be so good", "would you be so good as to", "could you be so good as to",
  "could you kindly", "could you so kindly", "i must say", "if i might say",
  "if i might be so bold", "if i may say so", "if i may be so bold as to say",
  "with permission", "with your permission", "with your consent", "with your leave",
  "if you'll permit me", "if you don't mind", "if you would not mind",
  "no offense", "no offense intended", "no disrespect", "no disrespect intended",
  "no harm meant", "no harm intended", "i mean no harm", "i intend no harm",
  "excuse me", "scuse me", "pardon me", "sorry", "very sorry", "i'm sorry",
  "i'm deeply sorry", "my apologies", "my sincere apologies", "my deepest apologies",
  "i apologize", "i sincerely apologize", "we apologize", "we sincerely apologize",
  "sincere apologies", "if i've offended", "if i've offended you", "if i've upset you",
  "i regret", "i deeply regret", "i do regret",
  "sir", "madam", "sir or madam", "your honor", "your majesty", "your grace",
  "your excellency", "your lordship", "your ladyship",
  "with pleasure", "with the greatest pleasure", "with all my pleasure",
  "with delight", "with great delight", "with all due courtesy", "with courtesy",
  "courteously", "civilly", "politely", "respectfully submitted",
  "most respectfully", "humbly", "humbly yours", "your humble servant",
  "humbly submitted", "if you so wish", "if you so choose", "if you so desire",
  "at your pleasure", "for your convenience", "for your benefit",
  "for your information", "fyi", "fyia", "fyii",
  "at your service", "at your disposal", "a pleasure", "my pleasure",
  "our pleasure", "my honor", "our honor",

  // Question softeners
  "could you", "would you", "would you mind", "can you", "can you please",
  "could you please", "would you please", "would you be so kind",
  "could you be so kind", "would you mind terribly", "i was wondering if",
  "i was thinking if", "i was wondering whether", "i was thinking whether",
  "would you be willing to", "would you mind if", "would you mind if i",
  "could you possibly", "would you possibly", "if you don't mind",
  "if you don't mind me", "if you would", "if you would be so kind",
  "if you would be willing", "if possible", "if you can", "if you could",
  "if you're able", "if you're available", "if it's not too much trouble",
  "if it's not too much bother", "if it's convenient", "if it's convenient for you",
  "if you have time", "if you have a moment", "if you have a spare moment",
  "if you have a few minutes", "when you get a chance", "when you have a moment",
  "when you have a minute", "when it's convenient", "at your convenience",
  "at your earliest convenience", "at your leisure", "whenever you can",
  "whenever you have time", "whenever it's convenient", "do you think you could",
  "do you think you might", "do you think you'd be willing", "would it be possible",
  "would it be possible for you", "would it be possible for you to",
  "would it be possible to", "is it possible", "is it possible for you",
  "is it possible to", "would there be any chance", "would there be any possibility",
  "would there be any way", "could there be any chance", "could there be any possibility",
  "would you have any chance", "would you have any way", "do you happen to",
  "could you happen to", "might you be able to", "might you possibly",
  "would you potentially", "could you potentially", "would you perhaps",
  "could you perhaps", "by any chance", "by any chance could you",
  "by any chance would you", "may i ask", "would you mind if i ask",
  "can i ask", "can i ask you", "can i ask you something", "might i ask",
  "would you mind me asking", "would you object to", "would you have any objection to",
  "would you consider", "would you be open to", "would you be willing",
  "would you entertain", "would you entertain the idea", "would you entertain the possibility",
  "would you give it a thought", "would you think about",
  "would you consider the possibility", "would you consider the idea",
  "would you give me a hand", "would you give me a hand with",
  "could you give me a hand", "could you give me a hand with",
  "would you mind giving me a hand", "would you mind helping",
  "could you help", "would you help", "would you be so good as to",
  "i would be grateful if", "i would be most grateful if",
  "i would appreciate it if", "i would greatly appreciate it if",
  "i would be thankful if", "we would be very grateful if",
  "we would appreciate it if", "would it be alright", "would it be alright with you",
  "would it be alright if", "is it alright", "is it alright if",
  "would it be conceivable", "would it be conceivable for", "would it be conceivable that",

  // Hedging and uncertainty
  "maybe", "perhaps", "possibly", "probably", "seems to", "appears to",
  "i think", "i believe", "i believe that", "i suspect", "i suspect that",
  "i suppose", "i suppose that", "i guess", "i imagined", "i figured",
  "i reckoned", "me thinks", "me thought", "in my opinion", "in my view",
  "in my humble opinion", "my opinion is", "my view is", "i'm of the opinion",
  "i'm of the view", "if you ask me", "from my perspective", "from my viewpoint",
  "from where i sit", "from what i gather", "from what i understand",
  "from what i can tell", "as far as i know", "as far as i'm aware",
  "as far as i can tell", "to my knowledge", "to the best of my knowledge",
  "if i recall correctly", "if my memory serves", "if memory serves me",
  "as far as i remember", "if i'm not mistaken", "if i'm not wrong",
  "correct me if i'm wrong", "i could be wrong", "i could be mistaken",
  "i could be off", "maybe i'm wrong", "maybe i'm mistaken", "maybe i'm off base",
  "perhaps i'm wrong", "i'm not sure", "i'm uncertain", "i'm not certain",
  "i'm doubtful", "i'm skeptical", "i have my doubts",
  "i'm not entirely convinced", "i'm not fully convinced",
  "seems like", "looks like", "sounds like", "feels like",
  "act as if", "for whatever reason", "for some reason", "for reasons unknown",
  "for no particular reason", "allegedly", "reportedly", "purportedly",
  "so called", "so-called", "speaking of", "in a sense", "sort of", "kind of",
  "semi", "partly", "partially", "to some extent", "to a degree",
  "to a certain extent", "to a certain degree", "more or less", "so to speak",
  "so to say", "in a manner of speaking", "if you like", "if you will",
  "if you prefer", "as it were", "as it happens", "as it may be",
  "apparently", "to all appearances", "by all appearances", "by all accounts",
  "by all indications", "at first glance", "at first blush",
  "roughly", "approximately", "somewhat", "rather", "instead of",
  "plausible", "conceivable", "hesitant", "reasonably",

  // Discourse markers
  "you know", "anyway", "anyhow", "by the way", "at any rate",
  "on second thought", "you see", "i mean", "as i was saying", "as i said",
  "back to", "back to business", "speaking of which", "getting back to",
  "in any case", "both sides", "besides", "then again", "on the other hand",
  "consequently", "accordingly", "admittedly", "as a matter of fact",
  "believe it or not", "come to think of it", "now that i think about it",
  "funny you should mention", "that reminds me", "along the same lines",
  "along similar lines", "as it turns out", "it turns out", "incidentally",
  "by the by", "the other day", "you know what", "i'll tell you what",
  "let me tell you", "what i mean is", "what i'm getting at",
  "what i'm trying to say", "what i'm saying is", "the thing is",
  "the thing about it", "here's the thing", "here's what i think",
  "across the board", "for the record", "on record", "i'm not gonna lie",
  "i'm being real", "i'm being honest", "real talk", "for real", "no cap",
  "no lie", "no joke", "it's not a joke", "speaking frankly", "frankly speaking",
  "to be frank", "if we're being honest", "if i'm being honest",
  "if we're being real", "one could say", "per se", "in effect",
  "in essence", "in short", "in brief", "in a nutshell", "in a word",
  "the way i see it", "the way i figure it", "once more", "yet once more",
  "quite a bit", "a fair amount", "quite a lot",
  "terribly", "awfully", "vastly", "wisely", "foolishly", "stupidly",
  "unfortunately", "funnily", "interestingly", "fascinatingly", "curiously",
  "remarkably", "strikingly", "completely", "entirely", "totally", "wholly",
  "unquestionably", "undoubtedly", "indubitably", "incontrovertibly",
  "period", "end of story", "odds are", "quite frankly", "truthfully",
  "without doubt", "without a doubt", "for sure", "for certain",

  // Obvious filler
  "to be honest", "if i'm honest", "now that i think about it",
  "come to think of it", "call me crazy", "i don't know",
  "how shall i put it", "quickly", "merely", "purely",
];

// Sort by length descending so longer phrases are matched first
PHRASE_PATTERNS.sort((a, b) => b.length - a.length);

// Vague pattern regexes (caught by characteristic, not explicit list)
const VAGUE_PATTERNS = [
  /\b(?:after a while|in a bit|earlier|later|soon|the other day)\b/gi,
  /\b(?:a bunch|a ton|loads of|tons of|quite a few|numerous)\b/gi,
  /\b(?:and stuff|and things|or whatever|or something|or anything|and whatnot|or whatnot)\b/gi,
  /\b(?:i guess you could say|how should i say|let me think|hmm|honestly)\b/gi,
  /\b(?:i'm no expert|i'm not sure but|correct me if wrong|not an expert but)\b/gi,
  /\b(?:moving on|to make long story short|bottom line|in conclusion)\b/gi,
  /\b(?:seriously|no kidding|you know what|get this)\b/gi,
  /\b(?:randomly|totally|completely by chance)\b/gi,
];

// Keywords that mark dismissible parenthetical asides
const ASIDE_KEYWORDS = [
  "ignore", "forget", "skip", "don't mind", "disregard",
  "not sure if", "not that", "not important", "random",
  "by the way", "tangent", "aside", "never mind"
];


// ─── Core Functions ────────────────────────────────────────────────────────

function normalizeText(text) {
  return text.replace(/\s+/g, " ").trim();
}

function removeOffTopicAsides(text) {
  const asides = [...text.matchAll(/\([^)]*\)/g)].map(m => m[0]);
  const removed = [];
  for (const aside of asides) {
    const lower = aside.toLowerCase();
    if (ASIDE_KEYWORDS.some(kw => lower.includes(kw))) {
      removed.push(aside);
      text = text.replace(aside, "");
    }
  }
  return { text: normalizeText(text), removed };
}

function removePhrasePatterns(text) {
  for (const phrase of PHRASE_PATTERNS) {
    // Escape special regex chars in the phrase
    const escaped = phrase.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
    const regex = new RegExp(`\\b${escaped}\\b`, "gi");
    text = text.replace(regex, " ");
  }
  return normalizeText(text);
}

function detectVaguePhrases(text) {
  const found = new Set();
  for (const pattern of VAGUE_PATTERNS) {
    pattern.lastIndex = 0;
    const matches = text.match(pattern) || [];
    matches.forEach(m => found.add(m));
  }
  return [...found];
}

function removeRedundantQualifiers(text) {
  // Repeated intensifiers: "very very" → "very"
  text = text.replace(/\b(very|really|so|quite|rather|extremely)\s+\1+\b/gi, "$1");
  // Multiple punctuation
  text = text.replace(/\?{2,}/g, "?");
  text = text.replace(/!{2,}/g, "!");
  text = text.replace(/\.{2,}/g, ".");
  // Empty parentheses/brackets
  text = text.replace(/\(\s*\)/g, "");
  text = text.replace(/\[\s*\]/g, "");
  // Double commas
  text = text.replace(/,\s*,/g, ",");
  text = text.replace(/^\s*,/, "");
  return normalizeText(text);
}

function smartDeduplicate(text) {
  let words = text.split(/\s+/);
  if (words.length === 0) return text;

  // Collapse adjacent duplicate single words
  const result = [words[0]];
  for (let i = 1; i < words.length; i++) {
    if (words[i].toLowerCase() !== words[i - 1].toLowerCase()) {
      result.push(words[i]);
    }
  }
  words = result;

  // Collapse repeated adjacent 2-3 word phrases
  for (const phraseLen of [3, 2]) {
    let i = 0;
    while (i + 2 * phraseLen <= words.length) {
      const a = words.slice(i, i + phraseLen).map(w => w.toLowerCase()).join(" ");
      const b = words.slice(i + phraseLen, i + 2 * phraseLen).map(w => w.toLowerCase()).join(" ");
      if (a === b) {
        words.splice(i + phraseLen, phraseLen);
      } else {
        i++;
      }
    }
  }

  return words.join(" ");
}

/**
 * Full optimizer — matches the Python optimizer.py behaviour.
 * @param {string} prompt - Input prompt
 * @param {boolean} aggressive - If true, applies extra removal passes
 * @param {boolean} removeAsides - If true, removes off-topic parenthetical asides
 * @returns {string} Optimized prompt
 */
function optimizePromptFull(prompt, aggressive = false, removeAsides = true) {
  prompt = normalizeText(prompt);

  // 1. Remove off-topic asides
  if (removeAsides) {
    const { text } = removeOffTopicAsides(prompt);
    prompt = text;
  }

  // 2. Remove explicit phrase patterns
  prompt = removePhrasePatterns(prompt);

  // 3. Remove vague phrases detected by pattern
  for (const vague of detectVaguePhrases(prompt)) {
    const escaped = vague.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
    prompt = prompt.replace(new RegExp(`\\b${escaped}\\b`, "gi"), " ");
  }

  // 4. Remove redundant qualifiers and cleanup
  prompt = removeRedundantQualifiers(prompt);

  // 5. Smart deduplication
  prompt = smartDeduplicate(prompt);

  // 6. Final normalize
  prompt = normalizeText(prompt);

  return prompt;
}

/**
 * Full prompt analysis with stats.
 */
function analyzePromptFull(prompt, removeAsides = true, aggressive = false) {
  const originalWords = prompt.split(/\s+/).filter(w => w).length;
  const vaguePhrases = detectVaguePhrases(prompt);
  const { removed: removedAsides } = removeOffTopicAsides(prompt);

  const optimized = optimizePromptFull(prompt, aggressive, removeAsides);
  const optimizedWords = optimized.split(/\s+/).filter(w => w).length;
  const reduction = originalWords > 0 ? ((originalWords - optimizedWords) / originalWords * 100) : 0;

  return {
    original_word_count: originalWords,
    optimized_word_count: optimizedWords,
    reduction_percentage: Math.round(reduction * 100) / 100,
    original: prompt,
    optimized,
    unknown_vague_phrases_detected: vaguePhrases,
    off_topic_asides: removedAsides,
  };
}

// ─── Public API (drop-in replacement for the old optimizer.js) ────────────

/**
 * Synchronous optimizer (legacy compat).
 */
function optimizePrompt(text) {
  return optimizePromptFull(text);
}

/**
 * Async optimizer — fully self-contained, no external services.
 * Returns same shape as before so server.js needs zero changes.
 */
async function optimizePromptAsync(text, aggressive = false) {
  const start = Date.now();
  const optimized = optimizePromptFull(text, aggressive);
  const originalWords = text.split(/\s+/).filter(w => w).length;
  const optimizedWords = optimized.split(/\s+/).filter(w => w).length;
  const reduction = originalWords > 0 ? ((originalWords - optimizedWords) / originalWords * 100) : 0;

  return {
    optimized,
    stats: {
      originalWords,
      optimizedWords,
      reductionPercent: Math.round(reduction * 100) / 100,
      processingTimeMs: Date.now() - start,
    },
    source: "js-advanced",
  };
}

/**
 * Analyzer — same as Python's analyze_prompt.
 */
async function callPythonAnalyzer(text) {
  return analyzePromptFull(text);
}

// Health checks — always return false since we no longer rely on Python services
async function isPythonOptimizerAvailable() { return false; }
async function isFastAPIOptimizerAvailable() { return false; }

// Keep these for any legacy import
async function callPythonOptimizer(text, aggressive = false) { return null; }
async function callFastAPIOptimizer(text, aggressive = false) { return null; }
function jsFallbackOptimize(text) { return optimizePromptFull(text); }

module.exports = {
  optimizePrompt,
  optimizePromptAsync,
  callPythonOptimizer,
  callFastAPIOptimizer,
  callPythonAnalyzer,
  isPythonOptimizerAvailable,
  isFastAPIOptimizerAvailable,
  jsFallbackOptimize,
};